lib/grpnames.gi

```diff
@@ -949,59 +949,81 @@ InstallMethod( DecompositionTypesOfGroup,
 ##
 #M  IsDihedralGroup( <G> ) . . . . . . . . . . . . . . . . . . generic method
 ##
-InstallMethod( IsDihedralGroup,
-               "generic method", true, [ IsGroup ], 0,
-
-  function ( G )
 
+DoComputeDihedralGenerators := function(G)
     local  Zn, G1, T, n, t, s, i;
 
-    if not IsFinite(G) then TryNextMethod(); fi;
-
-    if Size(G) mod 2 <> 0 then return false; fi;
+    if Size(G) mod 2 <> 0 then return fail; fi;
     n := Size(G)/2;
 
     # find a normal subgroup of G of type Zn
     if n mod 2 <> 0 then
       # G = < s, t | s^n = t^2 = 1, s^t = s^-1 >
       # ==> Comm(s, t) = s^-1 t s t = s^-2 ==> G' = < s^2 > = < s >
       Zn := DerivedSubgroup(G);
-      if not ( IsCyclic(Zn) and Size(Zn) = n ) then return false; fi;
+      if not ( IsCyclic(Zn) and Size(Zn) = n ) then return fail; fi;
     else # n mod 2 = 0
       # G = < s, t | s^n = t^2 = 1, s^t = s^-1 >
       # ==> Comm(s, t) = s^-1 t s t = s^-2 ==> G' = < s^2 >
       G1 := DerivedSubgroup(G);
-      if not ( IsCyclic(G1) and Size(G1) = n/2 ) then return false; fi;
+      if not ( IsCyclic(G1) and Size(G1) = n/2 ) then return fail; fi;
       # G/G1 = {1*G1, t*G1, s*G1, t*s*G1}
       T := RightTransversal(G,G1);
       i := 1;
       repeat
         Zn := ClosureGroup(G1,T[i]);
         i  := i + 1;
       until i > 4 or ( IsCyclic(Zn) and Size(Zn) = n );
-      if not ( IsCyclic(Zn) and Size(Zn) = n ) then return false; fi;
+      if not ( IsCyclic(Zn) and Size(Zn) = n ) then return fail; fi;
     fi; # now Zn is normal in G and Zn = < s | s^n = 1 >
 
     # choose t in G\Zn and check dihedral structure
     repeat t := Random(G); until not t in Zn;
     if not (Order(t) = 2 and ForAll(GeneratorsOfGroup(Zn),s->t*s*t*s=s^0))
-    then return false; fi;
+    then return fail; fi;
 
     # choose generator s of Zn
     repeat s := Random(Zn); until Order(s) = n;
-    SetDihedralGenerators(G,[t,s]);
+    return [t,s];
+end;
+
+InstallMethod( IsDihedralGroup,
+               "for a group",
+               true,
+               [ IsGroup and IsFinite ], 0,
+function(G)
+    local gens;
+
+    if not HasDihedralGenerators(G) then
```

Comment:
Options:  `InstallTrueMethod(IsDihedralGroup, HasDihedralGenerators);` in the `.gd` file, and then remove this `if` (the method can only be called if the dihedral generators are not (yet) known; though even if my claim was wrong, clearly it wouldn't be wrong to recompute them, only inefficient)

Of course the same could be done in the quaternion case.